### PR TITLE
define interface of data access object

### PR DESCRIPTION
**Pull request by ghost**
co-276

### CODE REVIEW

File: src/main/groovy/edu/oregonstate/mist/directoryapi/DirectoryEntityDAO.groovy [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import org.ldaptive.SearchOperation
+import org.ldaptive.SearchRequest
+import org.ldaptive.SearchResult
+
+/**
+ * Directory entity data access object.
+ */
+class DirectoryEntityDAO {
+    // FIXME: define properties in external file?
+    private final String LDAP_URL = 'ldap://client-ldap.onid.orst.edu'
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/directoryapi/DirectoryEntityDAO.groovy:16]: Yes, both of these strings should be coming from a configuration file. We don't want to be hard coding them like this.

File: src/main/groovy/edu/oregonstate/mist/directoryapi/DirectoryEntityDAO.groovy [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        List<DirectoryEntity> directoryEntityList = new ArrayList<DirectoryEntity>()
+        Connection connection = DefaultConnectionFactory.getConnection(LDAP_URL)
+        try {
+            connection.open()
+            SearchOperation operation = new SearchOperation(connection)
+            SearchRequest request = new SearchRequest(BASE_DN, filter)
+            Response<SearchResult> response = operation.execute(request)
+            SearchResult result = response.getResult()
+            for (LdapEntry entry : result.getEntries()) {
+                directoryEntityList.add(convert(entry))
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/directoryapi/DirectoryEntityDAO.groovy:74]: There should probably be a catch statement that goes along with this try statement, we can then log any errors that occur.

File: src/main/groovy/edu/oregonstate/mist/directoryapi/DirectoryEntityDAO.groovy [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * Searches LDAP directory for filter string.
+     *
+     * @param filter
+     * @return directoryEntityList
+     */
+    private List<DirectoryEntity> searchLDAP(String filter) {
+        List<DirectoryEntity> directoryEntityList = new ArrayList<DirectoryEntity>()
+        Connection connection = DefaultConnectionFactory.getConnection(LDAP_URL)
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/directoryapi/DirectoryEntityDAO.groovy:66]: Could line 61 cause an exception or error? I'm wonder if it should be within the try block.

File: src/main/groovy/edu/oregonstate/mist/directoryapi/DirectoryEntityDAO.groovy [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        directoryEntityList
+    }
+
+    /**
+     * Converts LdapEntry to DirectoryEntity.
+     *
+     * @param ldapEntry
+     * @return DirectoryEntity
+     */
+    private DirectoryEntity convert(LdapEntry ldapEntry) {
```
Comment (floppyears) [inline src/main/groovy/edu/oregonstate/mist/directoryapi/DirectoryEntityDAO.groovy]: This function will be a great candidate for several unit tests. The connection to ldap can be mocked in other test methods as needed.

**Comment by browtayl** [comment]
@floppyears this file is under active development and i am the current assignee. the

**Comment by ghost** [comment]
@floppyears this file is under active development and i am the current assignee. the inline comments are reminders to myself. the code for accessing ldap is directly from the library's official documentation.

**Comment by ghost** [comment]
library documentation link: http://www.ldaptive.org/docs/guide/operations/search.html

**Pull request merged**